### ISSUE TITLE
OAuth2: Unregistered Scopes and/or additional data via Access Token

### ISSUE THREAD

**Issue by DonMartin76** [Kong/kong#1759]
I have the following situation: When identifying a user using our own authentication system (SAML based) I also get information on the user being authorized (licenses) to access the API or not. I would like to pass on this information with the upstream call, enriched via the access token.

We want to use Kong to secure a content API which delivers different results depending on which content products (identified by product IDs) the user has licensed.

A possibility to do this would be using OAuth2 scopes. The problem I have now is quite stupid: We  literally have thousands of product IDs, and each one would need to be registered with the API's OAuth2 Scopes. And they change once in a while, meaning I would have to keep the API definition updated via some kind of integration.

My question now is: Could I solve this in a different way? The obvious way is to tell the API implementation to go ask the license server itself, but it sort of feels stupid as I do have the information at hand in the Authorization Server.

Other ways I see which would be nice (or at least would solve my problem):
- Allow the `oauth2` plugin to allow arbitrary scopes when authorizing; i.e. allow an access token to be created for any list of scopes, even if the plugin of the API does not have scopes defined (this is what I mean with "unregistered scopes").
- Allow the authorize end point to not only pass on `X-Authenticated-UserId`, but additionally arbitrary other headers, e.g. in my case `X-Authenticated-Products` (this is what I mean with "additional data via Access Token").

These both "solutions" would probably mean changes to the OAuth2 Plugin. It would definitely increase the usability for my use cases if one of those were implemented.

I could also encode the information inside the `X-Authenticated-UserId` header, but that feels strange.

**Comment by hanshuebner** [comment]
@DonMartin76 Is this issue still relevant or did you find a solution?

**Comment by DonMartin76** [comment]
We found a bunch of workarounds, mostly we resorted to encoding more information into the `X-Authenticated-UserId` header. This kind of feels wrong, but does the work. We also implemented that abovementioned integration for one use case, where we periodically poll one service and update the OAuth2 scopes of the API with an updated list of product information.

So, strictly speaking we're probably not needing this anymore.
